In [16]:
import pandas as pd

In [17]:
df = pd.read_csv('train.csv')

In [18]:
df

datetime  season  holiday  workingday  weather   temp  \
0      2011-01-01 00:00:00       1        0           0        1   9.84   
1      2011-01-01 01:00:00       1        0           0        1   9.02   
2      2011-01-01 02:00:00       1        0           0        1   9.02   
3      2011-01-01 03:00:00       1        0           0        1   9.84   
4      2011-01-01 04:00:00       1        0           0        1   9.84   
...                    ...     ...      ...         ...      ...    ...   
10881  2012-12-19 19:00:00       4        0           1        1  15.58   
10882  2012-12-19 20:00:00       4        0           1        1  14.76   
10883  2012-12-19 21:00:00       4        0           1        1  13.94   
10884  2012-12-19 22:00:00       4        0           1        1  13.94   
10885  2012-12-19 23:00:00       4        0           1        1  13.12   

        atemp  humidity  windspeed  casual  registered  count  
0      14.395        81     0.0000       3          13     16  
1      13.635        80     0.0000       8          32     40  
2      13.635        80     0.0000       5          27     32  
3      14.395        75     0.0000       3          10     13  
4      14.395        75     0.0000       0           1      1  
...       ...       ...        ...     ...         ...    ...  
10881  19.695        50    26.0027       7         329    336  
10882  17.425        57    15.0013      10         231    241  
10883  15.910        61    15.0013       4         164    168  
10884  17.425        61     6.0032      12         117    129  
10885  16.665        66     8.9981       4          84     88  

[10886 rows x 12 columns]

In [19]:
X = df[['temp', 'humidity']]
y = df['count']

In [20]:
X.head()

temp  humidity
0  9.84        81
1  9.02        80
2  9.02        80
3  9.84        75
4  9.84        75

In [21]:
# polynomial feature engineering
from sklearn.preprocessing import PolynomialFeatures # transformative model
# regularization
from sklearn.linear_model import Ridge # predictive model 

In [22]:
from sklearn.pipeline import make_pipeline
# metamodel
# puts different models into a pipeline
# to avoid the doubling of fit transform for train and test data

In [23]:
from sklearn.ensemble import VotingRegressor # bonus combination model you can try out

In [24]:
m = make_pipeline(PolynomialFeatures(), Ridge())
m

Pipeline(memory=None,
         steps=[('polynomialfeatures',
                 PolynomialFeatures(degree=2, include_bias=True,
                                    interaction_only=False, order='C')),
                ('ridge',
                 Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                       max_iter=None, normalize=False, random_state=None,
                       solver='auto', tol=0.001))],
         verbose=False)

# FunctionTransformer
you can also take your own custom feature engineering function and wrap it into a sklearn object called a FunctionTransformer, , so that you can also include it in the pipeline

In [25]:
m.fit(X, y) #polynomialfeatures.fit_transform(), followed by ridge.fit()

Pipeline(memory=None,
         steps=[('polynomialfeatures',
                 PolynomialFeatures(degree=2, include_bias=True,
                                    interaction_only=False, order='C')),
                ('ridge',
                 Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                       max_iter=None, normalize=False, random_state=None,
                       solver='auto', tol=0.001))],
         verbose=False)

How to choose the optimal combination of hyperparameters?

In [26]:
from sklearn.model_selection import GridSearchCV

In [27]:
PARAM_GRID = {'polynomialfeatures__degree':[1, 2, 3, 4, 5, 6, 7, 8],
             'ridge__alpha':[0.5, 0.6, 1.0, 2.0, 5.0]}

grid = GridSearchCV(m, param_grid=PARAM_GRID)
# deault scoring method is R2
# but we can customize it to the log ...

In [29]:
grid.fit(X, y)

/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=8.95603e-20): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=8.63373e-20): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=8.58035e-20): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=8.01999e-20): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=8.33319e-20): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib

/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.02592e-27): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=9.74733e-28): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.04058e-27): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.38373e-27): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.30865e-27): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib

/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.10317e-30): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.0356e-30): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=9.73541e-31): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.06146e-30): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning: Ill-conditioned matrix (rcond=1.31982e-35): result may not be accurate.
  overwrite_a=True).T
/Users/lara/anaconda3/lib/

GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('polynomialfeatures',
                                        PolynomialFeatures(degree=2,
                                                           include_bias=True,
                                                           interaction_only=False,
                                                           order='C')),
                                       ('ridge',
                                        Ridge(alpha=1.0, copy_X=True,
                                              fit_intercept=True, max_iter=None,
                                              normalize=False,
                                              random_state=None, solver='auto',
                                              tol=0.001))],
                                verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'polynomialfeatures__degree': [1

In [30]:
grid.best_params_

{'polynomialfeatures__degree': 4, 'ridge__alpha': 5.0}

In [32]:
pd.DataFrame(grid.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.004073      0.000468         0.001753        0.000221   
1        0.004468      0.000931         0.001512        0.000178   
2        0.003368      0.000237         0.001270        0.000122   
3        0.003339      0.000340         0.001206        0.000153   
4        0.003174      0.000113         0.001354        0.000176   
5        0.004185      0.001124         0.001394        0.000216   
6        0.003449      0.000249         0.001188        0.000022   
7        0.003743      0.000426         0.001241        0.000074   
8        0.003532      0.000219         0.001194        0.000021   
9        0.003462      0.000173         0.001277        0.000088   
10       0.004077      0.000631         0.001620        0.000295   
11       0.003830      0.000257         0.001450        0.000215   
12       0.003905      0.000256         0.001438        0.000254   
13       0.003777      0.000140         0.001325        0.000076   
14       0.004121      0.000496         0.001503        0.000205   
15       0.004745      0.000407         0.001741        0.000202   
16       0.004666      0.000578         0.001762        0.000169   
17       0.005119      0.000272         0.001658        0.000067   
18       0.004667      0.000419         0.001899        0.000214   
19       0.005315      0.000599         0.001678        0.000112   
20       0.006367      0.000598         0.002259        0.000375   
21       0.005170      0.000548         0.001906        0.000207   
22       0.004992      0.000326         0.001839        0.000081   
23       0.005220      0.000421         0.001828        0.000111   
24       0.005075      0.000492         0.001885        0.000185   
25       0.006022      0.000758         0.002140        0.000197   
26       0.006176      0.000557         0.002098        0.000348   
27       0.005616      0.000460         0.002011        0.000199   
28       0.005442      0.000142         0.001937        0.000113   
29       0.005598      0.000143         0.001977        0.000104   
30       0.007452      0.001105         0.002269        0.000157   
31       0.006274      0.000599         0.002312        0.000263   
32       0.007954      0.001164         0.002475        0.000407   
33       0.006743      0.000227         0.002110        0.000182   
34       0.006804      0.000339         0.002527        0.000431   
35       0.007881      0.001171         0.002565        0.000221   
36       0.006867      0.000379         0.002468        0.000207   
37       0.007162      0.000463         0.002187        0.000177   
38       0.007590      0.000973         0.002542        0.000252   
39       0.007366      0.000488         0.002228        0.000208   

   param_polynomialfeatures__degree param_ridge__alpha  \
0                                 1                0.5   
1                                 1                0.6   
2                                 1                  1   
3                                 1                  2   
4                                 1                  5   
5                                 2                0.5   
6                                 2                0.6   
7                                 2                  1   
8                                 2                  2   
9                                 2                  5   
10                                3                0.5   
11                                3                0.6   
12                                3                  1   
13                                3                  2   
14                                3                  5   
15                                4                0.5   
16                                4                0.6   
17                                4                  1   
18                                4                  2   
19                                4                  5 

In [33]:
final_model = grid.best_estimator_

In [34]:
final_model.predict(X_test)
# score and predict on the testing data AT THE VERY END

NameError: name 'X_test' is not defined

A Polynomial of Degree 4 on 2 features ('temp', 'humdity'):
    - temp^2*hum
    - temp^3*hum
    - temp^4*hum
    - temp^2*hum^2
    - temp^2*hum^2

### other things you can put into a pipeline:
1. Custom Feature Engineering (e.g. see: FunctionTransformer)
1.1 
2. Scaling
3. PCA
4. Model